src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -159,7 +159,10 @@ internal JsonConverter GetConverterFromMember(Type? parentClassType, Type proper
         [RequiresUnreferencedCode("Getting a converter for a type may require reflection which depends on unreferenced code.")]
         public JsonConverter GetConverter(Type typeToConvert!!)
         {
-            RootBuiltInConverters();
+            if (!IsInitializedForReflectionSerializer)
+            {
+                InitializeForReflectionSerializer();
+            }
```

Comment:
I don't think calling this method should root the reflection-based metadata gathering code paths, potentially masking future bugs with source-gen.

Reply:
Thanks. What is the distinction between rooting the reflection-based metadata gathering paths and rooting the built-in converter factories that depend on said reflection-based metadata gathering paths? Under the current setup an options instance not configured for reflection will eagerly return converters that do require reflection.

Reply:
Based on your remarks on this PR https://github.com/dotnet/runtime/pull/51897 I'm guessing that this is only meant as an entrypoint for reflection-based serialization anyways.

Reply:
> Under the current setup an options instance not configured for reflection will eagerly return converters that do require reflection.

Hmm what do you mean? The source generator doesn't use this method `GetConverter(Type)`, so we left the reflection to avoid breaking folks that used it pre .NET 6.0.